lib/Dialect/ESI/ESIServices.cpp

```diff
@@ -109,8 +113,133 @@ static LogicalResult instantiateCosimEndpointOps(ServiceImplementReqOp req,
   return success();
 }
 
+// Generator for "sv_mem" implementation type. Emits SV ops for an unpacked
+// array, hopefully inferred as a memory to the SV compiler.
+static LogicalResult
+instantiateSystemVerilogMemory(ServiceImplementReqOp req,
+                               ServiceDeclOpInterface decl) {
+  ImplicitLocOpBuilder b(req.getLoc(), req);
+  BackedgeBuilder bb(b, req.getLoc());
+
+  RandomAccessMemoryDeclOp ramDecl =
+      dyn_cast<RandomAccessMemoryDeclOp>(decl.getOperation());
+  if (!ramDecl)
+    return req.emitOpError("'sv_mem' implementation type can only be used to "
+                           "implement RandomAccessMemory declarations");
```

Comment:
Nit: missing test... but i assume it would be redundant seeing as the general service decl-impl logic is already tested.

Reply:
Strictly speaking, the IR allows custom declarations to use the sv_mem implementation, though that could change with #3942 . There should be a more principled way of handling this for builtin implementations.